internal/parser/strconv_test.go

```diff
@@ -0,0 +1,167 @@
+package parser_test
+
+import (
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/influxdata/flux/ast"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/influxdata/flux/internal/parser"
+)
+
+func TestParseDuration(t *testing.T) {
+	testCases := []struct {
+		testName string
+		duration string
+		want     []ast.Duration
+		err      error
+	}{
+		{
+			testName: "All durations",
+			duration: "1y3mo2w1d4h1m30s1ms2µs70ns",
+			want: []ast.Duration{
+				{
+					Magnitude: 1,
+					Unit:      "y",
+				},
+				{
+					Magnitude: 3,
+					Unit:      "mo",
+				},
+				{
+					Magnitude: 2,
+					Unit:      "w",
+				},
+				{
+					Magnitude: 1,
+					Unit:      "d",
+				},
+				{
+					Magnitude: 4,
+					Unit:      "h",
+				},
+				{
+					Magnitude: 1,
+					Unit:      "m",
+				},
+				{
+					Magnitude: 30,
+					Unit:      "s",
+				},
+				{
+					Magnitude: 1,
+					Unit:      "ms",
+				},
+				{
+					Magnitude: 2,
+					Unit:      "us",
+				},
+				{
+					Magnitude: 70,
+					Unit:      "ns",
+				},
+			},
+		},
+		{
+			testName: "Leading zero durations",
+			duration: "01y03mo02w01d04h01m03s01ms02µs07ns",
+			want: []ast.Duration{
+				{
+					Magnitude: 1,
+					Unit:      "y",
+				},
+				{
+					Magnitude: 3,
+					Unit:      "mo",
+				},
+				{
+					Magnitude: 2,
+					Unit:      "w",
+				},
+				{
+					Magnitude: 1,
+					Unit:      "d",
+				},
+				{
+					Magnitude: 4,
+					Unit:      "h",
+				},
+				{
+					Magnitude: 1,
+					Unit:      "m",
+				},
+				{
+					Magnitude: 3,
+					Unit:      "s",
+				},
+				{
+					Magnitude: 1,
+					Unit:      "ms",
+				},
+				{
+					Magnitude: 2,
+					Unit:      "us",
+				},
+				{
+					Magnitude: 7,
+					Unit:      "ns",
+				},
+			},
+		},
+		{
+			testName: "Many leading zeros duration",
+			duration: "000000000001234d",
```

Comment:
Yeah, the WG decided to allow an arbitrary number of leading `0`s instead of playing with crafting error messages for restricting to just 1 `0\d+` or `0` or `00`